### PR TITLE
ESQL: Add nulls support to Categorize

### PR DESCRIPTION
**Pull request by ivancea**
Handle nulls and empty strings (Which resolve to null) on Categorize grouping function.

Also, implement `seenGroupIds()`, which would fail some queries with nulls otherwise.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizeRawBlockHash.java [hunk 1]
```diff
@@ -95,7 +95,8 @@ public IntBlock eval(int positionCount, BytesRefBlock vBlock) {
                 BytesRef vScratch = new BytesRef();
                 for (int p = 0; p < positionCount; p++) {
                     if (vBlock.isNull(p)) {
-                        result.appendNull();
+                        seenNull = true;
```
[the remark this answers is not in the document]
Comment (ivancea) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizeRawBlockHash.java:98]: :')

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizedIntermediateBlockHash.java [hunk 2]
```diff
@@ -40,9 +40,22 @@ public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
...
+                addInput.add(0, newIds);
+            }
+            return;
+        }
+
         Map<Integer, Integer> idMap = readIntermediate(categorizerState.getBytesRef(0, new BytesRef()));
         try (IntBlock.Builder newIdsBuilder = blockFactory.newIntBlockBuilder(idMap.size())) {
-            for (int i = 0; i < idMap.size(); i++) {
+            int categoryIdEnd = idMap.size() + 1;
+            if (idMap.containsKey(0)) {
```
Comment (jan-elastic) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizedIntermediateBlockHash.java:54]: I find this easier to read:
```
fromId = idMap.containsKey(0) ? 0 : 1
toId = fromId + idMap.size()
for (id = fromId ... toId)
    newIdsBuilder.appendInt(idMap.get(id));
```

**Review by jan-elastic — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/categorize.csv-spec [hunk 3]
```diff
@@ -92,20 +92,33 @@ COUNT():long | category:keyword
...
+required_capability: categorize_v2
+
+FROM employees
+  | STATS SUM(languages) BY category=CATEGORIZE(job_positions)
+  | SORT category DESC
+  | LIMIT 2
+;
+
+SUM(languages):long | category:keyword
+                 27 | null
```
[the remark this answers is not in the document]
Comment (ivancea) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/categorize.csv-spec:120]: A null, as we understand it. It's a special case while reading this table.
So to the question, "what if we wanted to represent a "null" string? I have no idea! :hehe:

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizedIntermediateBlockHash.java (hunk 2 above)
Comment (ivancea) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizedIntermediateBlockHash.java]: Oh, far more elegant. Pushed!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java [hunk 4]
```diff
@@ -76,24 +88,39 @@ public final ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue target
...
+                result.appendNull();
+                for (SerializableTokenListCategory category : categorizer.toCategoriesById()) {
+                    scratch.copyChars(category.getRegex());
+                    result.appendBytesRef(scratch.get());
+                    scratch.clear();
+                }
+                return result.build();
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (jan-elastic) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:122]: OK, thx for explaining

**Review by alex-spies — COMMENTED** [review]
I know I'm late to the party, but LGTM and thanks @ivancea !

**Pull request merged**